zrml/hybrid-router/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+                    ]
+                }
+            }
+        }
+
+        /// Fills the order from the Automated Market Maker (AMM) if it exists.
+        /// The order is filled until the `price_limit` is reached.
+        ///
+        /// # Arguments
+        ///
+        /// * `tx_type` - The type of transaction (Buy or Sell).
+        /// * `who` - The account ID of the user performing the transaction.
+        /// * `market_id` - The ID of the market.
+        /// * `asset` - The asset to be traded.
+        /// * `amount_in` - The amount to be traded.
+        /// * `price_limit` - The maximum or minimum price at which the trade can be executed.
+        fn maybe_fill_from_amm(
+            tx_type: TxType,
+            who: &AccountIdOf<T>,
+            market_id: MarketIdOf<T>,
+            asset: AssetOf<T>,
+            amount_in: BalanceOf<T>,
+            price_limit: BalanceOf<T>,
+        ) -> Result<(BalanceOf<T>, Option<AmmTradeOf<T>>), DispatchError> {
+            if !T::Amm::pool_exists(market_id) {
+                return Ok((amount_in, None));
+            }
+
+            let spot_price = T::Amm::get_spot_price(market_id, asset)?;
+
+            let amm_amount_in = match tx_type {
+                TxType::Buy => {
+                    if spot_price >= price_limit {
+                        return Ok((amount_in, None));
+                    }
+                    T::Amm::calculate_buy_amount_until(market_id, asset, price_limit)?
+                }
+                TxType::Sell => {
+                    if spot_price <= price_limit {
+                        return Ok((amount_in, None));
+                    }
+                    T::Amm::calculate_sell_amount_until(market_id, asset, price_limit)?
+                }
+            };
+
+            let amm_amount_in = amm_amount_in.min(amount_in);
+
+            if amm_amount_in.is_zero() {
+                return Ok((amount_in, None));
+            }
+
+            let amm_trade_info = Self::handle_amm_trade(
+                tx_type,
+                who.clone(),
+                market_id,
+                asset,
+                amm_amount_in,
+                BalanceOf::<T>::zero(),
+            )?;
+
+            Ok((amount_in.checked_sub_res(&amm_amount_in)?, amm_trade_info))
+        }
+
+        fn handle_amm_trade(
+            tx_type: TxType,
+            who: AccountIdOf<T>,
+            market_id: MarketIdOf<T>,
+            asset: AssetOf<T>,
+            amount_in: BalanceOf<T>,
+            min_amount_out: BalanceOf<T>,
+        ) -> Result<Option<AmmTradeOf<T>>, DispatchError> {
+            match tx_type {
+                TxType::Buy => {
+                    match T::Amm::buy(who, market_id, asset, amount_in, min_amount_out) {
+                        Ok(amm_trade) => Ok(Some(amm_trade)),
+                        Err(ApiError::SoftFailure(AmmSoftFail::Numerical)) => Ok(None),
```

Comment:
The soft failure execution path appears to be untested. If you go with a super small buy amount for a low price outcome, you should be able to reliably create a soft failure.

Reply:
d5a36ba563b7ae32411f69c107f0aca67e3b5da9

Reply:
Looks good, although I think we need to work on the logic that sets the price of the limit order (not in this PR though). I think if the purchase doesn't move the needle on the AMM, it shouldn't place a limit order at a significantly higher price.